### PR TITLE
(#135) Setup library module

### PR DESCRIPTION
**Pull request by pankaj-ag**

### CODE REVIEW

**Comment by pankaj-ag** [comment]
@rathorevaibhav I have set up the basic structure of KnowledgeCafe feature. I will code it further after your code review on the current code. 

Features I build so far - 
1- Added new tab in the home page. 
2-User can click a image of the book and see the information.

File: app/Http/Controllers/KnowledgeCafe/Library/BookController.php [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    public function update(Request $request, $id)
+    {
+        //
+    }
+
+    /**
+     * Remove the specified resource from storage.
+     *
+     * @param  int  $id
+     * @return \Illuminate\Http\Response
```
[the remark this answers is not in the document]
Comment (pankaj-ag) [reply app/Http/Controllers/KnowledgeCafe/Library/BookController.php]: done.

File: app/Http/Controllers/KnowledgeCafe/Library/BookController.php [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+
+     /**
+     * Fetch the book info.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return \Illuminate\Http\Response
+     */
+    public function fetchBookInfo(Request $request) {
+        $method = $request->input('add_method');
```
[the remark this answers is not in the document]
Comment (pankaj-ag) [reply app/Http/Controllers/KnowledgeCafe/Library/BookController.php]: done.

File: app/Http/Controllers/KnowledgeCafe/Library/BookController.php [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            $file = $request->file('book_image');
+            $ISBNNumber = BookServices::getISBN($file);
+
+            if(!$ISBNNumber || strlen($ISBNNumber) < 13) {
+                return json_encode(['view'=> null, 'error' => true, 'message' => 'Could not fetch valid ISBN : '. $ISBNNumber]);
+            }
+            $book= BookServices::getBookDetails($ISBNNumber);
+        } 
+
+        if($method ==='from_isbn') {
```
[the remark this answers is not in the document]
Comment (pankaj-ag) [reply app/Http/Controllers/KnowledgeCafe/Library/BookController.php]: done.

File: app/Http/Controllers/KnowledgeCafe/Library/BookController.php [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+     */
+    public function fetchBookInfo(Request $request) {
+        $method = $request->input('add_method');
+
+        if($method === 'from_image') {
+            $file = $request->file('book_image');
+            $ISBNNumber = BookServices::getISBN($file);
+
+            if(!$ISBNNumber || strlen($ISBNNumber) < 13) {
+                return json_encode(['view'=> null, 'error' => true, 'message' => 'Could not fetch valid ISBN : '. $ISBNNumber]);
```
[the remark this answers is not in the document]
Comment (pankaj-ag) [reply app/Http/Controllers/KnowledgeCafe/Library/BookController.php]: done.

File: app/Http/Controllers/KnowledgeCafe/Library/BookController.php [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            $book= BookServices::getBookDetails($ISBNNumber);
+        } 
+
+        if($method ==='from_isbn') {
+            $ISBNNumber = $request->input('isbn');
+            $book = BookServices::getBookDetails($ISBNNumber);
+        }  
+
+        if(!isset($book['items'])) {
+            return json_encode(['view'=> null, 'error' => true, 'message' => 'Invalid ISBN : '. $ISBNNumber]);
```
[the remark this answers is not in the document]
Comment (pankaj-ag) [reply app/Http/Controllers/KnowledgeCafe/Library/BookController.php]: done.

File: app/Http/Controllers/KnowledgeCafe/Library/BookController.php [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                    'view'=> null, 
+                    'error' => true, 
+                    'message' => 'Could not fetch valid ISBN : '. $ISBNNumber
+                ]);
+            }
+
+            $book= BookServices::getBookDetails($ISBNNumber);
+        } else if($method ==='from_isbn') {
+            $ISBNNumber = $request->input('isbn');
+            $book = BookServices::getBookDetails($ISBNNumber);
```
Comment (bohraji) [inline app/Http/Controllers/KnowledgeCafe/Library/BookController.php]: Please add check to verify if the info on the book with the ISBN number is stored in the application database. The will prevent additional api call's for the same book.

File: app/Services/BookServices.php [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+        $apiKey = env('GOOGLE_VISION_API_KEY');
+        $vision = new Vision( $apiKey, [new Feature(Feature::TEXT_DETECTION, 100)]);
+        $response = $vision->request(new LocalImage($file->path()));
+        $faces = $response->getTextAnnotations();
+        $description = '';
+        $currentText = '';
+
+        foreach ($faces as $face) {
+            $faceDescription = $face->getDescription();
```
Comment (bohraji) [inline app/Services/BookServices.php]: Unused variable

File: app/Http/Controllers/KnowledgeCafe/Library/BookController.php [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+            if(!$ISBNNumber || strlen($ISBNNumber) < 13) {
+                return response()->json([
+                    'view'=> null, 
+                    'error' => true, 
+                    'message' => 'Could not fetch valid ISBN : '. $ISBNNumber
+                ]);
+            }
+
+            $book= BookServices::getBookDetails($ISBNNumber);
```
Comment (bohraji) [inline app/Http/Controllers/KnowledgeCafe/Library/BookController.php]: As the final info on the book is being fetched via ISBN, can we have local var `$ISBN` which stores the input from request. If the request is containing $file, then fetch the number and call the book service at a common point.

`function() {
$isbn = $request->isbn;
if(method == file) {
 //get $isbn
}
//check ISBN Format
//get book info via Service
}`

File: app/Http/Controllers/KnowledgeCafe/Library/BookController.php (hunk 6 above)
Comment (pankaj-ag) [reply app/Http/Controllers/KnowledgeCafe/Library/BookController.php]: Agreed. 
Tracking this in (#141).

File: app/Http/Controllers/KnowledgeCafe/Library/BookController.php (hunk 8 above)
Comment (pankaj-ag) [reply app/Http/Controllers/KnowledgeCafe/Library/BookController.php]: Yes, @bohraji.
I just want to leave this package a bit decoupled. That's why I added that check here
https://github.com/ColoredCow/employee-portal/pull/136/files/c46b4a9c5215ed8edd39fd2ecb72b15e4fdbbfe4#diff-be0f344d20d66c0af83aa2ea0b578a4bR98

File: app/Services/BookServices.php (hunk 7 above)
Comment (bohraji) [reply app/Services/BookServices.php]: Fixed

**Pull request merged**